README.md
<p align="center">
  <img width="320" height="167" src="https://raw.githubusercontent.com/PortBlueSky/thread-puddle/75aec302f67b4542cc1577a2803d0ab38046b93f/assets/tp-logo.png">
</p>

# thread-puddle

## Turn any module into a worker thread

[![Master Push](https://github.com/PortBlueSky/thread-puddle/actions/workflows/master-push.yml/badge.svg)](https://github.com/PortBlueSky/thread-puddle/actions/workflows/master-push.yml) [![npm version](https://badge.fury.io/js/thread-puddle.svg)](https://badge.fury.io/js/thread-puddle) [![Coverage Status](https://coveralls.io/repos/github/PortBlueSky/thread-puddle/badge.svg?branch=master)](https://coveralls.io/github/PortBlueSky/thread-puddle?branch=master)

A library to pool Node.js [worker threads](https://nodejs.org/dist/latest-v13.x/docs/api/worker_threads.html), automatically exposing exported module methods using [Proxy Objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) for easy bi-directional communication between main and worker thread.

__+ Full TypeScript Support__ (optionally using [ts-node](https://github.com/TypeStrong/ts-node) in development)  
__+ Only one dependency__ (the `debug` package)  
__+ Callbacks__ (workers can call back functions on the main thread)


### Installation

```bash
npm install thread-puddle
# or
yarn add thread-puddle
```

### Usage Example

```ts
// examples/ts-class/worker.ts
export class MyWorker {
  say(): void {
    console.log('Hello!')
  }
}

export default new MyWorker()
```

```ts
// examples/ts-class/main.ts
import { createThreadPool } from '../../lib'
import { MyWorker } from './worker'

async function start () {
  const worker = await createThreadPool<MyWorker>('./worker')
  await worker.say() // -> "Hello!"
  
  worker.pool.terminate()
}

start()
```

This and more examples can be found in the `examples` directory.


## Typing

`createThreadPool` uses TypeScript [generics](https://www.typescriptlang.org/docs/handbook/generics.html) and [other](https://www.typescriptlang.org/docs/handbook/utility-types.html#picktk) [advanced](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterst) [features](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypet) to give you the type of how your worker module will actually be exposed to the main thread.

__TL;DR__: The captured type for your thread will be modified to only return async methods and be extended with the pool interface.

Example:

```ts
interface Calculations {
  crunchNumbers(data: number[]): number;
}

// Using as worker type:
const worker = createThreadPool<Calculations>('./calc-worker')

// Will basically become:
interface Calculations {
  crunchNumbers(data: number[]): Promise<number>;
  pool: PoolInterface
}
```


## API

### `async createThreadPool<T>(workerPath, [options])`

Creates a pool of workers and waits until all workers are ready to call methods on, then returns a Proxy Object which will forward method calls to the worker.

Arguments:

- `workerPath` - an absolute or relative path to a module.
- `options` - optional settings:
  - `size`- the number of worker threads to be created in the pool for the given module. Defaults to `1`.
  - `workerOptions` - will be used as options for the [worker thread constructor](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_new_worker_filename_options). Defaults to `{}`.
  - `startupTimeout` - if a worker thread cannot be started within this timout in milliseconds, the pool creation will fail and reject with a timout error. Defaults to `30000`.
  - `typecheck` - In development `ts-node` is used, which by default runs `transpile-only` mode. To get type checks, set this to `true`.
  - `maxQueueSize` - When filled up with waiting calls, will reject all subsequent calls, until emptied to below max again. Defaults to `1000`
  - `autoRefill` - Automatically fills up the pool with workers until `size` is reached when workers unexpectedly exit. Defaults to `false`.

If the pool size is `> 1`, method calls will be forwarded to the next available worker. If all workers are busy, the method calls will be queued. A worker will handle one method call at any time only.

### `async worker.[method]([arguments])`

On the Proxy Object returned from `createThreadPool`, you can call any method which is exported from the `workerPath` module. **All method calls are async and return a Promise**, no matter if the module method is async or not. The promise will be resolved with the return value of the method in the worker module. If the method call fails in the worker thread, the Promise will be rejected with that error and the original stack trace. If a method is not available on the worker, the Promise will be rejected with an error.

- `method` - Must match a method name exported from the worker module.
- `arguments` - Arbitrary number of arguments forwarded to the method call in the worker thread.

`Arguments` are transferred to the worker thread via [`postMessage`](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_port_postmessage_value_transferlist), compatible with the [HTML structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm). If you want to move arguments of type `ArrayBuffer` or `MessageChannel` instead of copying them, you can use the [`withTransfer`](#withtransfervalue-transferlist) helper.

`Arguments` can contain a __`function`__ (__callback__) in the first level, which will be callable from the thread via a reference, but be executed on the main thread.
__However__, the result of the callback is currently __not__ transfered back to the thread. It is possible to implement a worker thread with an `EventEmitter` though, see the [eventemitter example](./examples/ts-eventemitter/). Callback errors will throw an `unhandled exception` if no error listener is attached to the pool `callback:error` event like `worker.pool.on('callback:error', () => ...)`.

```ts
// examples/ts-callbacks/worker.ts
export class MyWorker {
  callMe(callback: (msg: string) => void): void {
    callback('Hello!')
  }
}

export default new MyWorker()
```

```ts
// examples/ts-callbacks/main.ts
import { createThreadPool } from '../../lib'
import { MyWorker } from './worker'

async function start () {
  const worker = await createThreadPool<MyWorker>('./worker')
  const callback = (msg: string) => console.log(msg)
  await worker.callMe(callback) 
  
  worker.pool.terminate()
}

start()
```

### `async worker.all.[method]([arguments])`

Will call the given `method` on all workers, as soon as they become available. Returns a list of the results from all workers, like `Promise.all`. Otherwise behaves the same as `worker.[method]([arguments])`.

### `worker.pool.terminate()`

Terminates the pool and all worker threads in it. Trying to call methods in the pool afterwards, will result in an rejection.

### `worker.pool.refill()`

Fills up the pool with workers until `size` is reached. Can be used to manually decide wether to refill or terminate.

### `worker.pool.drain()`

Waits until all calls are handled and threads are idle, then terminates all threads.

### `worker.pool.size`

The number of worker threads in the pool.

### `worker.pool.isTerminated`

Wether or not the pool was terminated. 

### `withTransfer(value, [transferList])`

This helper can be used **bi-directional**, to transfer values to a worker thread as method call argument(s), or to transfer return values from a worker thread method.

Arguments:

- `value` - Any value which contains transferrables (`ArrayBuffer` or `MessageChannel`).
- `transferList` - If `value` itself is not a [typed array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays), the list may specify any transferrable contained in `value`.

Example method arguments:

```js
// Main thread
const buf = Buffer.from('my main buffer value')
const value = {
  buf,
  anotherValue: true
}

worker.customMethod(withTransfer(value, [buf])) // <--
```

Example return value:

```js
// Worker thread
module.exports = {
  customMethod() {
    const buf = Buffer.from('my thread buffer value')
    const value = {
      buf,
      anotherValue: true
    }
    return withTransfer(value, [buf]) // <--
  }
}
```

### Pool Events

#### `error`

Forwards errors that are emitted for a specific worker in the pool. When a worker thread errors, it is terminated. The pool will spawn another worker, unless it is terminated.

#### `exit`

Forwards exit events that are emitted for a specific worker in the pool, adding the `threadId` as second argument.

## Debug

Thread Puddle has only one dependency: [`debug`](https://www.npmjs.com/package/debug)

Namespaces:

- `puddle:*` - all debug logs for the module
- `puddle:master` - only logs from the pool controller
- `puddle:thread:*` - logs for all threads only
- `puddle:thread:[id]` - logs for a specific thread

Nested Namespaces:

- `puddle:parent:[id]:master`
- `puddle:parent:[id]:thread:[id]`

The debug method with the correct namespace is exported from the `thread-puddle` entry point.

---
&copy; 2022 Sebastian Herrlinger

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## Version 0.4.0
  - Basic Callback Functions #21
    - Workers can now implement EventEmitter behaviour
  - Worker paths can now be relative
  - ts-node is now `transpile-only` per default #21
  - Allows chaining (resolves to the proxy when the worker returns itself)
  - Upgrade all dev dependencies
  - Improved typing
  - Max Queue Size #29
    - Subsequent calls will be rejected when max queue size is reached,
    until the queue is emptied below max again.
  - Refill
    - automatically refill pool when workers exit via `autoRefill: true` pool option
    - manually refill pool whenever via `worker.pool.refill()`
  - Drain (graceful shutdown)
    - Wait until all calls are done and workers are idle, then terminate, via `worker.pool.drain()`

## Version 0.3.2
  - Stricter and fixed types #16

## Version 0.3.1
  - Make ts-node optional #7 (@spmiller)
  - Bump some dev dependency versions

## Version 0.3.0

- TypeScript support :tada:
  Building a TypeScript worker on-the-fly with ts-node,
  generic `createThreadPool<OwnType>` pool creation
  with generated type from a custom worker type (promise wrapped methods).

## Version 0.2.1, 0.2.2

- Bump dev dependency versions for security update

## Version 0.2.0

- Puddle internal thread id is exposed to worker via `const { threadId } = require('thread-puddle')`.
- The module now exports `createThreadPool` as a pool constructor.
- Loads ECMAScript Modules (ES6 Modules) as workers for node version >= 13.
- Has now only one way to access the pool interface: `workerProxy.pool`
- Emits worker events on the `workerProxy.pool` interface
- Emits an `exit` event when a worker exits, `code` and `threadId`


src/index.ts
import path from 'path'
import { EventEmitter } from 'events'
import createDebug from 'debug'
import getCallsites from './utils/callsites'

export { withTransfer } from './Transferable'
import { isMainThread, WorkerOptions } from 'worker_threads'
import hasTSNode from './utils/has-ts-node'
import { ThreadId } from './types/general'
import { createSequence } from './utils/sequence'
import { WorkerThread } from './WorkerThread'
import { RequestQueue } from './components/request-queue'

const { threadId: dynamicThreadId, debug: dynamicDebug } = require('./export-bridge')
export const threadId = dynamicThreadId

const workerThreadIdSequence = createSequence({ start: 1 })

let debugOut = createDebug('puddle:master')
export let debug = debugOut
if (!isMainThread) {
  debugOut = createDebug(`puddle:parent:${threadId}:master`)
  debug = dynamicDebug
}

export interface ThreadRequest {
  resolve(thread: WorkerThread): void
  reject(error: Error): void
}

export type ThreadPoolOptions = {
  size?: number
  typecheck?: boolean
  workerOptions?: WorkerOptions
  startupTimeout?: number
  maxQueueSize?: number
  autoRefill?: boolean
}

export interface BaseWorker {
  pool: PoolInterface
}

export interface PoolInterface extends EventEmitter {
  terminate(): void
  refill(): void
  drain(shouldTerminate?: boolean): Promise<void>
  size: number
  isTerminated: boolean
  threads: ReadonlyMap<ThreadId, WorkerThread>
}

type ProxyWorkerTarget = Record<string, any>

type FilterType<Base, Condition> = Pick<Base, {
  [Key in keyof Base]: Base[Key] extends Condition ? Key : never
}[keyof Base]>;

type TypeWithMethods = Record<string | number | symbol,  (...a: any) => any | Promise<any>>
export type AsyncMethod = (...param: any) => Promise<any>

// TODO: 
// Should reject sync callback methods in parameters
// (or re-type them to async?)
type WrapReturnType<Base extends TypeWithMethods> = {
  [Key in keyof Base]: Base[Key] extends AsyncMethod
    ? Base[Key] 
    : (...a: Parameters<Base[Key]>) => Promise<ReturnType<Base[Key]>>;
};

// @ts-ignore
// TODO: Fix usage of interface vs. type
// Even though this complains, the type is inferred from the template correctly,
// working for classes/interfaces and types
type FilterAndWrap<Base> = WrapReturnType<FilterType<Required<Base>, Function>> 

export type WrapWorkerType<Base> = FilterAndWrap<Base> & BaseWorker & { all: FilterAndWrap<Base> }

export async function createThreadPool<T> (workerPath: string, {
  size = 1,
  workerOptions = {},
  startupTimeout = 30000,
  typecheck = false,
  maxQueueSize = 1000,
  autoRefill = false
}: ThreadPoolOptions = {}) {
  // Validate Options
  if (maxQueueSize < size) {
    throw new Error('maxQueueSize needs to be at least the number of workers in the pool')
  }

  debugOut('carving out a puddle...')

  // Resolve relative worker path
  let resolvedWorkerPath = workerPath

  if (!path.isAbsolute(workerPath)) {
    let callsites = getCallsites()
    if (callsites) {
      callsites = callsites.filter((cs) => cs.getFileName())
      const callerPath = callsites[1].getFileName()
      const { dir: basePath} = path.parse(callerPath!)
      resolvedWorkerPath = path.resolve(basePath, workerPath)
    }
  }

  const { ext: resolvedWorkerExtension } = path.parse(require.resolve(resolvedWorkerPath))

  const bridgeWorkerPath = path.resolve(__dirname, 'worker')
  const { ext: bridgeWorkerExtension } = path.parse(require.resolve(bridgeWorkerPath))
  let workerString = `require('${bridgeWorkerPath}')`

  if (hasTSNode() && [bridgeWorkerExtension, resolvedWorkerExtension].includes('.ts')) {
    if (typecheck) {
      workerString = `require('ts-node').register()\n${workerString}`
    } else {
      workerString = `require('ts-node/register/transpile-only')\n${workerString}`
    }
  }
  
  // TODO: Automatically infer types from worker path if not given
  // const implicitWorkerType = await import('./__tests__/workers/basic');
  // type WorkerType = unknown extends T
  //   ? typeof implicitWorkerType
  //   : T;
  // -> See: https://github.com/microsoft/TypeScript/issues/31090
  type WorkerType = T
  
  type TargetWorkerType = BaseWorker & { all: FilterAndWrap<WorkerType> }
  type ExtendedWorkerType = TargetWorkerType & FilterAndWrap<WorkerType>

  // Based on: https://github.com/Microsoft/TypeScript/issues/20846#issuecomment-582183737
  interface PoolProxyConstructor {
    new <T, H extends object, K extends BaseWorker>(target: T, handler: ProxyHandler<H>): K
  }

  interface PoolProxyAllConstructor {
    new <T, H extends object, K extends WorkerType>(target: T, handler: ProxyHandler<H>): K
  }

  const threads = new Map<ThreadId, WorkerThread>()
  const availableThreads: WorkerThread[] = []
  const threadRequests = new RequestQueue()

  // Holds the number of queued direct calls via worker.all
  let directCallQueueSize = 0;
  
  let isTerminated = false

  const allWorkersTarget = {}
  const puddleInterface = new EventEmitter()

  const removeThread = ({ id }: WorkerThread) => {
    threads.delete(id)
    availableThreads.splice(availableThreads.findIndex(thread => (thread.id === id)), 1)
  }

  const drain = async (shouldTerminate: boolean = true) => {
    if (threadRequests.size() > 0 || Array.from(threads.values()).some((thread) => thread.callQueue.length > 0)) {
      await new Promise<void>((resolve) => puddleInterface.on('ready', () => {
        if (threads.size === availableThreads.length) {
          resolve()
        }
      }))
    }
    
    if (shouldTerminate) {
      terminate()
    }
  }

  const refill = () => {
    const diff = size - threads.size

    for (let i = 0; i < diff; i++) {
      createThread()
    }
  }

  const createThread = () => {
    debugOut('creating worker thread')

    const thread = new WorkerThread(workerThreadIdSequence.next(), threadId, debugOut, workerString, resolvedWorkerPath, workerOptions)
    threads.set(thread.id, thread)

    debugOut('creates worker thread %s', thread.id)

    thread.on('ready', (id) => {
      if (threadRequests.hasPending()) {
        const request = threadRequests.next()
        request!.resolve(thread)
        return
      }
      
      availableThreads.push(thread)
      
      puddleInterface.emit('ready', id)
    })

    thread.on('message', (msg, id) => {
      puddleInterface.emit('thread:message', msg, id)
    })

    // Forward callback errors to worker.pool interface
    thread.on('callback:error', (err, id) => {
      if (puddleInterface.listenerCount('callback:error') > 0) {
        puddleInterface.emit('callback:error', err, id)
        return
      }
      throw err
    })

    thread.on('startup-error', (err) => {
      // TODO: There might be more than one request
      // -> startup error should be handled elsewhere
      if (threadRequests.hasPending()) {
        const request = threadRequests.next()
        request!.reject(err)
      }
    })

    thread.on('exit', (code, id) => {
      removeThread(thread)
      puddleInterface.emit('exit', code, id)

      if (!isTerminated && autoRefill) {
        refill()
      }
      
      if (threads.size === 0) {
        terminate()

        const err = new Error('All workers exited before resolving (use an error event handler or DEBUG=puddle:*)')
        threadRequests.rejectAll(err)
      }
    })

    thread.on('error', (err, id) => {
      if (puddleInterface.listenerCount('error') > 0) {
        puddleInterface.emit('error', err, id)
      } else {
        terminate()
      }
    })
  }

  debugOut('filling puddle with thread liquid...')
  refill()

  const getAvailableThread = () => new Promise<WorkerThread>((resolve, reject) => {
    if (isTerminated) {
      return reject(new Error('Worker pool already terminated.'))
    }

    if (availableThreads.length > 0) {
      const thread = availableThreads.shift()!
      debugOut('Resolving available worker %d', thread!.id)
      return resolve(thread)
    }

    if (threadRequests.size() + directCallQueueSize >= maxQueueSize) {
      return reject(new Error('Max thread queue size reached'))
    }

    const threadRequest: ThreadRequest = { resolve, reject }
    threadRequests.push(threadRequest)
    debugOut('Added worker request')
  })

  const terminate = () => {
    if (isTerminated) {
      return
    }
    
    debugOut('pulling the plug on the puddle...')
    
    isTerminated = true
    
    threads.forEach((thread) => {
      thread.terminate()
    })
  }

  await Promise.all(Array.from(threads.values()).map((thread) => new Promise<void>((resolve, reject) => {
    const timeout = setTimeout(() => {
      terminate()
      reject(new Error(`Worker ${thread.id} initialization timed out`))
    }, startupTimeout)
    const threadRequest: ThreadRequest = {
      resolve: (thread) => {
        debugOut('worker %d ready', thread.id)

        clearTimeout(timeout)
        availableThreads.push(thread)
        resolve()
      },
      reject: (err) => {
        debugOut('worker %d startup failed', thread.id)

        clearTimeout(timeout)
        terminate()
        reject(err)
      }
    }

    threadRequests.push(threadRequest)
  })))

  debugOut('puddle filled, happy splashing!')

  // TODO: Metrics
  // Optionally gather metrics for method calls,
  // to get a metrics object like:
  // { function1: { roundTrip: { avg: 25, median: 23, max: 934 }, calls: 1564 } }

  Object.assign(puddleInterface, {
    terminate,
    refill,
    drain
  })
  Object.defineProperties(puddleInterface, {
    size: {
      get: () => threads.size
    },
    isTerminated: {
      get: () => isTerminated
    },
    threads: {
      get: () => threads
    }
  })

  const PoolProxy = Proxy as PoolProxyConstructor
  const PoolProxyAll = Proxy as PoolProxyAllConstructor
  const allWorkersHandler = {
    get: (target: ProxyWorkerTarget, key: string) => {
      // NOTE: If the proxy is returned from an async function,
      // the engine checks if it is a thenable by checking existence of a then method
      if (key === 'then') {
        return undefined
      }

      if (threadRequests.size() + directCallQueueSize >= maxQueueSize) {
        return () => Promise.reject(new Error('Max thread queue size reached'))
      }

      availableThreads.splice(0)
      return (...args: any[]) => Promise.all(
        Array.from(threads.values()).map(thread => new Promise((resolve, reject) => {
          directCallQueueSize += 1
          thread.callOnThread(key, args, (value: any) => {
            directCallQueueSize -= 1
            resolve(value)
          }, (err) => {
            directCallQueueSize -= 1
            reject(err)
          })
        }))
      )
    }
  }

  const allWorkersInterface: WorkerType = new PoolProxyAll<
    typeof allWorkersTarget, 
    typeof allWorkersHandler, 
    WorkerType
  >(allWorkersTarget, allWorkersHandler)

  const target: TargetWorkerType = {
    pool: puddleInterface as PoolInterface,
    all: allWorkersInterface as FilterAndWrap<WorkerType>
  }

  // Intermediate, so the proxy can return itself
  let proxy: ExtendedWorkerType
  
  const handler = {
    get: (proxyTarget: ProxyWorkerTarget, key: string) => {
      // NOTE: If the proxy is returned from an async function,
      // the engine checks if it is a thenable by checking existence of a then method
      if (key === 'then') {
        return undefined
      }

      if (key === 'pool' || key === 'all') {
        return proxyTarget[key]
      }

      return async (...args: any[]) => {
        const result = await new Promise((resolve, reject) => {
          getAvailableThread()
            .then((thread: WorkerThread) => thread.callOnThread(key, args, resolve, reject))
            .catch(err => reject(err))
        })
        
        if (result === '__THIS__') {
          return proxy
        }

        return result
      }
    }
  }

  proxy = new PoolProxy<typeof target, typeof handler, ExtendedWorkerType> (target, handler)
  return proxy
}
